controllers/imageclusterinstall_controller.go

```diff
@@ -207,21 +210,32 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 
 	// in case there is no bmh we should set requirements met condition to true with image ready message
 	// in case bmh was set we will set this condition after host validations
-	if ici.Status.BareMetalHostRef == nil {
+	if ici.Spec.BareMetalHostRef == nil {
 		if err := r.setImageReadyCondition(ctx, ici, nil, imageUrl); err != nil {
 			log.WithError(err).Error("failed to update ImageClusterInstall status")
 			return ctrl.Result{}, err
 		}
 	}
 
 	if ici.Status.BareMetalHostRef != nil && !v1alpha1.BMHRefsMatch(ici.Spec.BareMetalHostRef, ici.Status.BareMetalHostRef) {
-		if err := r.removeBMHImage(ctx, ici.Status.BareMetalHostRef); client.IgnoreNotFound(err) != nil {
+		if _, err := r.removeBMHImage(ctx, ici.Status.BareMetalHostRef); client.IgnoreNotFound(err) != nil {
 			log.WithError(err).Errorf("failed to remove image from BareMetalHost %s/%s", ici.Status.BareMetalHostRef.Namespace, ici.Status.BareMetalHostRef.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
 	if bmh != nil {
+		// in case image data was changed and bmh has image url configured
+		// we should remove image from it in order to invalidate ironic cache
+		if bmh.Spec.Image != nil && updated {
+			r.Log.Info("Image data was changed, removing image from BareMetalHost")
+			removed, err := r.removeBMHImage(ctx, ici.Spec.BareMetalHostRef)
+			if err != nil || removed {
+				// in case of error log will be printed by reconcile loop
```

Comment:
We are in `Reconcile` currently, who is printing this log?

Reply:
In case error occur manager prints the error